README.md
# default-backend

```bash
docker pull ysicing/defaultbackend-v2
# 替换ingress default backend
```

## 使用

```yaml
# kubectl get cm/nginx-ingress-nginx-controller  -n kube-system -o yaml
apiVersion: v1
data:
  access-log-path: /var/log/nginx/nginx_access.log
  compute-full-forwarded-for: "true"
  custom-http-errors: 404,503
  error-log-path: /var/log/nginx/nginx_error.log
  forwarded-for-header: X-Forwarded-For
  generate-request-id: "true"
  keep-alive-requests: "10000"
  log-format-upstream: $remote_addr - $remote_user [$time_iso8601] $msec "$request"
    $status $body_bytes_sent "$http_referer" "$http_user_agent" $request_length $request_time
    [$proxy_upstream_name] [$proxy_alternative_upstream_name] [$upstream_addr] [$upstream_response_length]
    [$upstream_response_time] [$upstream_status] $req_id
  max-worker-connections: "65536"
  proxy-body-size: 50m
  upstream-keepalive-connections: "200"
  use-forwarded-headers: "true"
kind: ConfigMap
```

## v1版本参考

[defaultbackend-v1](https://github.com/ysicing/dockerfiles/tree/v1.0/defaultbackend)


CHANGELOG.md
# Changelog

- 0.2.0
  - 优化页面

- 0.1.1
  - 启动输出版本信息

- 0.1.0
  - 优化部分代码

- 0.0.8
  - 优化部分代码

- 0.0.6
  - 优化部分代码

- 0.0.4
  - 优化部分代码

- 0.0.2
  - 添加健康检查

- 0.0.1
  - 初始版本


pkg/k8s/k8s.go
package k8s

import (
	"fmt"

	"github.com/ysicing/default-backend/internal/kube"
	"k8s.io/client-go/informers"
	appsinformer "k8s.io/client-go/informers/apps/v1"
	ingressv1 "k8s.io/client-go/informers/networking/v1"
	"k8s.io/client-go/kubernetes"
	"k8s.io/client-go/tools/cache"
)

var Client kubernetes.Interface
var IngressInformer ingressv1.IngressInformer

func init() {
	var err error
	kubecfg := &kube.ClientConfig{}
	Client, err = kube.New(kubecfg)
	if err != nil {
		panic(err)
	}
}

type Controlller struct {
	informerFactory    informers.SharedInformerFactory
	ingressInformer    ingressv1.IngressInformer
	deploymentInformer appsinformer.DeploymentInformer
}

func (c *Controlller) create(obj interface{}) {

}

func (c *Controlller) delete(obj interface{}) {

}

func (c *Controlller) update(oldobj, newobj interface{}) {

}

func (c *Controlller) Run(stopCh chan struct{}) error {
	c.informerFactory.Start(stopCh)
	if !cache.WaitForCacheSync(stopCh, c.ingressInformer.Informer().HasSynced) {
		return fmt.Errorf("failed to sync caches")
	}
	return nil
}

func NewControlller(i informers.SharedInformerFactory) *Controlller {
	IngressInformer = i.Networking().V1().Ingresses()
	c := &Controlller{
		informerFactory:    i,
		ingressInformer:    IngressInformer,
		deploymentInformer: i.Apps().V1().Deployments(),
	}
	c.ingressInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
		AddFunc:    c.create,
		DeleteFunc: c.delete,
		UpdateFunc: c.update,
	})
	return c
}


pkg/k8s/ingress.go
package k8s

import (
	"context"

	"github.com/ergoapi/util/exmap"
	"github.com/ergoapi/util/ptr"
	"github.com/ergoapi/zlog"
	"github.com/ysicing/default-backend/internal/cache"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"k8s.io/apimachinery/pkg/labels"
)

var (
	NotExistCode = 0
	ExistRunning = 1
	ExistStart   = 2
	Exist        = 99
)

type Meta struct {
	Host      string
	Name      string
	Namespace string
	Appid     string
	Status    int
}

func GetIngress(host string) *Meta {
	zlog.Debug("check host: %s", host)
	var m Meta
	ings, _ := IngressInformer.Lister().List(labels.Everything())
	for _, ing := range ings {
		if ing.Name == host {
			m.Host = host
			m.Appid = exmap.GetLabelValue(ing.Labels, "k8s.easycorp.work/appid")
			m.Name = exmap.GetLabelValue(ing.Labels, "k8s.easycorp.work/name")
			m.Namespace = ing.Namespace
			if len(m.Appid) == 0 || len(m.Name) == 0 {
				m.Status = NotExistCode
			} else {
				m.Status = Exist
			}
			return &m
		}
	}
	m.Status = NotExistCode
	return &m
}

func StartDeploy(host string) int {
	m := GetIngress(host)
	if m.Status == NotExistCode {
		return m.Status
	}
	dg, err := Client.AppsV1().Deployments(m.Namespace).Get(context.TODO(), m.Name, metav1.GetOptions{})
	if err != nil {
		return NotExistCode
	}
	replicas := *dg.Spec.Replicas
	if replicas > 0 {
		if cache.Check(host) {
			return ExistStart
		}
		return ExistRunning
	}
	dg.Spec.Replicas = ptr.Int32Ptr(1)
	_, err = Client.AppsV1().Deployments(m.Namespace).Update(context.TODO(), dg, metav1.UpdateOptions{})
	if err != nil {
		zlog.Error("update deployment: %v", err)
	}
	cache.Set(host)
	return ExistStart
}


internal/kube/kube.go
package kube

import (
	"log"
	"os"
	"path/filepath"

	"github.com/ergoapi/util/zos"
	"k8s.io/client-go/kubernetes"
	"k8s.io/client-go/rest"
	"k8s.io/client-go/tools/clientcmd"
)

type ClientConfig struct {
	QPS   float32
	Burst int
}

// New returns a kubernetes client.
// It tries first with in-cluster config, if it fails it will try with out-of-cluster config.
func New(cc *ClientConfig) (client kubernetes.Interface, err error) {
	client, err = NewInCluster(cc)
	if err == nil {
		log.Println("kubernetes client is in-cluster")
		return
	}

	dir, err := os.UserHomeDir()
	if err != nil {
		return
	}

	dir = filepath.Join(dir, ".kube", "config")

	if zos.IsMacOS() {
		dir = "/Users/ysicing/.kube/sh-k3s.kubeconfig"
	}

	client, err = NewFromConfig(cc, dir)
	if err != nil {
		return
	}
	log.Println("kubernetes client is kubeconfig")
	return
}

// NewFromConfig returns a new out-of-cluster kubernetes client.
func NewFromConfig(cc *ClientConfig, path string) (client kubernetes.Interface, err error) {
	// use the current context in kubeconfig
	config, err := clientcmd.BuildConfigFromFlags("", path)
	if err != nil {
		return
	}

	cc.apply(config)

	// create the clientset
	clientset, err := kubernetes.NewForConfig(config)
	if err != nil {
		return
	}

	client = clientset

	return
}

// NewInCluster returns a new in-cluster kubernetes client.
func NewInCluster(cc *ClientConfig) (client kubernetes.Interface, err error) {
	// creates the in-cluster config
	config, err := rest.InClusterConfig()
	if err != nil {
		return
	}

	cc.apply(config)

	// creates the clientset
	clientset, err := kubernetes.NewForConfig(config)
	if err != nil {
		return
	}

	client = clientset

	return
}

func (cc *ClientConfig) apply(config *rest.Config) {
	if cc.QPS > 0.0 {
		config.QPS = cc.QPS // the default is rest.DefaultQPS which is 5.0
	}

	if cc.Burst > 0 {
		config.Burst = cc.Burst // the default is rest.DefaultBurst which is 10
	}
}


pkg/server/server.go
package server

import (
	"context"
	"fmt"
	"html/template"
	"net/http"
	"os"
	"strings"
	"time"

	"github.com/ergoapi/exgin"
	"github.com/ergoapi/util/environ"
	"github.com/ergoapi/util/version"
	_ "github.com/ergoapi/util/version/prometheus"
	"github.com/ergoapi/zlog"
	"github.com/gin-gonic/gin"
	"github.com/prometheus/client_golang/prometheus/promhttp"
	"k8s.io/apimachinery/pkg/util/validation"
	"k8s.io/client-go/informers"

	"github.com/ysicing/default-backend/pkg/k8s"
	"github.com/ysicing/default-backend/pkg/templates"
)

func fake(c *gin.Context) bool {
	ua := c.Request.UserAgent()
	ua = strings.ToLower(ua)
	path := c.Request.URL.Path
	if strings.Contains(ua, "python") || strings.Contains(ua, "nss") || strings.Contains(ua, "nmap") || strings.Contains(ua, "censys") {
		return true
	}
	if strings.Contains(ua, "bot") || strings.Contains(ua, "bytespider") || strings.Contains(ua, "android") || strings.Contains(ua, "windows") || strings.Contains(ua, "iphone") || strings.Contains(ua, "ipad") || strings.Contains(ua, "nss") || strings.Contains(ua, "nmap") || strings.Contains(ua, "censys") {
		return true
	}
	if strings.Contains(path, ".php") || strings.Contains(path, "wp") {
		return true
	}
	return false
}

func fakehost(host string) bool {
	defaultHost := environ.GetEnv("DOMAIN_SUFFIX", "ysicing.net")
	return !strings.HasSuffix(host, defaultHost)
}

func Index(c *gin.Context) {
	host := c.Request.Host
	ip := exgin.RealIP(c)
	if len(validation.IsDNS1123Subdomain(host)) != 0 {
		c.HTML(403, "ip.html", gin.H{
			"host": host,
			"ip":   ip,
		})
		return
	}

	s := k8s.StartDeploy(host)
	if fake(c) || fakehost(host) || s == k8s.NotExistCode {
		c.HTML(403, "4xx.html", gin.H{
			"host": host,
			"ip":   ip,
		})
		return
	}

	c.HTML(200, "5xx.html", gin.H{
		"host": host,
		"ip":   ip,
	})
}

func Serve(ctx context.Context) error {
	stopChan := make(chan struct{})
	factory := informers.NewSharedInformerFactory(k8s.Client, time.Minute)
	controller := k8s.NewControlller(factory)
	controller.Run(stopChan)
	g := exgin.Init(environ.GetEnv("ENVTYPE", "prod") == "prod")
	g.Use(exgin.ExCors())
	g.Use(exgin.ExLog("/healthz", "/metrics"))
	g.Use(exgin.ExRecovery())
	tpls := template.Must(template.New("").ParseFS(templates.FS, "pages/*.html"))
	g.SetHTMLTemplate(tpls)
	g.GET("/", Index)
	g.GET("/healthz", func(c *gin.Context) {
		exgin.GinsData(c, map[string]string{
			"healthz": "healthz",
		}, nil)
	})
	g.GET("/metrics", gin.WrapH(promhttp.Handler()))
	g.GET("/rv", func(c *gin.Context) {
		v := version.Get()
		exgin.GinsData(c, map[string]string{
			"builddate": v.BuildDate,
			"gitcommit": v.GitCommit,
			"version":   v.GitVersion,
		}, nil)
	})
	g.NoMethod(func(c *gin.Context) {
		msg := fmt.Sprintf("not found: %v", c.Request.Method)
		exgin.GinsAbortWithCode(c, 404, msg)
	})
	g.NoRoute(Index)
	addr := "0.0.0.0:65001"
	srv := &http.Server{
		Addr:    addr,
		Handler: g,
	}
	go func() {
		defer close(stopChan)
		<-ctx.Done()
		ctx, cancel := context.WithTimeout(context.TODO(), time.Second*5)
		defer cancel()
		if err := srv.Shutdown(ctx); err != nil {
			zlog.Error("Failed to stop server, error: %s", err)
		}
		zlog.Info("server exited.")
	}()
	zlog.Info("http listen to %v, pid is %v, version: %v", addr, os.Getpid(), version.GetShortString())
	if err := srv.ListenAndServe(); err != nil && err != http.ErrServerClosed {
		zlog.Error("Failed to start http server, error: %s", err)
		return err
	}

	<-stopChan

	return nil
}
